examples/config.c

```diff
@@ -0,0 +1,304 @@
+		struct device_config *dc;
+		bool found = false;
+
+		wl_list_for_each(dc, &config->devices, link) {
+			if (strcmp(dc->name, device_name) == 0) {
+				found = true;
+				break;
+			}
+		}
+
+		if (!found) {
+			dc = calloc(1, sizeof(struct device_config));
+			dc->name = strdup(device_name);
+			wl_list_insert(&config->devices, &dc->link);
+		}
+
+		if (strcmp(name, "map-to-output") == 0) {
+			if (dc->mapped_output) {
+				free(dc->mapped_output);
+			}
+			dc->mapped_output = strdup(value);
+		} else if (strcmp(name, "geometry") == 0) {
+			if (dc->mapped_geo) {
+				free(dc->mapped_geo);
+			}
+			dc->mapped_geo = parse_geometry(value);
+		} else {
+			wlr_log(L_ERROR, "got unknown device config: %s", name);
+		}
+	} else {
+		wlr_log(L_ERROR, "got unknown config section: %s", section);
+	}
+
+	return 1;
+}
+
+struct example_config *parse_args(int argc, char *argv[]) {
+	struct example_config *config = calloc(1, sizeof(struct example_config));
+	wl_list_init(&config->outputs);
+	wl_list_init(&config->devices);
+
+	int c;
+	while ((c = getopt(argc, argv, "C:h")) != -1) {
+		switch (c) {
+		case 'C':
+			config->config_path = strdup(optarg);
+			break;
+		case 'h':
+		case '?':
+			usage(argv[0], c != 'h');
+		}
+	}
+
+	if (!config->config_path) {
+		// get the config path from the current directory
+		char cwd[1024];
+		if (getcwd(cwd, sizeof(cwd)) != NULL) {
+			char buf[1024];
+			sprintf(buf, "%s/%s", cwd, "wlr-example.ini");
+			config->config_path = strdup(buf);
+		} else {
+			wlr_log(L_ERROR, "could not get cwd");
+			exit(1);
+		}
+	}
+
+	int result = ini_parse(config->config_path, config_ini_handler, config);
+
+	if (result == -1) {
+		wlr_log(L_ERROR, "Could not find config file at %s", config->config_path);
+		exit(1);
+	} else if (result == -2) {
+		wlr_log(L_ERROR, "Could not allocate memory to parse config file");
+		exit(1);
+	} else if (result != 0) {
+		wlr_log(L_ERROR, "Could not parse config file");
+		exit(1);
+	}
+
+	return config;
+}
+
+void example_config_destroy(struct example_config *config) {
+	struct output_config *oc, *otmp = NULL;
+	wl_list_for_each_safe(oc, otmp, &config->outputs, link) {
+		free(oc->name);
+		free(oc);
+	}
+
+	struct device_config *dc, *dtmp = NULL;
+	wl_list_for_each_safe(dc, dtmp, &config->devices, link) {
+		free(dc->name);
+		if (dc->mapped_output) {
+			free(dc->mapped_output);
+		}
+		if (dc->mapped_geo) {
+			free(dc->mapped_geo);
+		}
+		free(dc);
+	}
+
+	if (config->config_path) {
+		free(config->config_path);
+	}
+	if (config->cursor.mapped_output) {
+		free(config->cursor.mapped_output);
+	}
+	if (config->cursor.mapped_geo) {
+		free(config->cursor.mapped_geo);
+	}
+	free(config);
+}
+
+struct wlr_output_layout *configure_layout(struct example_config *config, struct wl_list *outputs) {
+	struct wlr_output_layout *layout = wlr_output_layout_init();
+	int max_x = INT_MIN;
+	int max_x_y = INT_MIN; // y value for the max_x output
```

Comment:
(nitpick) y_at_max_x? If you feel the need to comment a variable name, there might be a better fit -- not sure mine is :P